### DISCUSSION TITLE
How can I get the request payload?

### DISCUSSION THREAD

**Discussion by jacksonlancaster** [scottoffen/grapevine#146] [General]
IHttpContext context.Request.Payload was available in 4.2 version.
But it seems to have been removed.
How do I go about getting the request payload?

I tried this code below. Sometimes it works. Sometimes it doesn't return any data (even when the content-length is greater than 0.

using var reader = new StreamReader(context.Request.InputStream, context.Request.ContentEncoding);
var payload = await reader.ReadToEndAsync();

**Comment by scottoffen** [comment]
`HttpRequest.InputStream` is the correct property to get the payload from. However, because this is a wrapper class for [`HttpListenerRequest`](https://learn.microsoft.com/en-us/dotnet/api/system.net.httplistenerrequest?view=net-8.0), the underlying stream is read-once, and the position cannot be reset.

In v4.x and lower, I would read this as a string into the Payload property, but this was problematic for the same reason that it would be problematic to read the `InputStream` into a `MemoryStream` - when your dealing with large inputs, this could cause resource issues.

v5.x doesn't have a better solution for this currently. If you need the payload to be available during multiple routes, I would recommend deserializing it and adding to the `context.Locals` collection.

**Reply by cuican6** [reply]
Is it feasible to create a body attribute that reads data from the stream when the body is first retrieved, and then retrieves the body using the read cache in the body.
